File: bindings/cpp/samples/monitor_recv_sample.cpp
/* SPDX-License-Identifier: MPL-2.0 */

#include "sample_common.hpp"

int main ()
{
    zlink::context_t ctx;
    zlink::pair_socket_t server (ctx);
    zlink::pair_socket_t client (ctx);
    zlink::socket_monitor_t server_monitor =
      server.monitor_open (zlink::monitor_event::connection_ready);
    zlink::socket_monitor_t client_monitor =
      client.monitor_open (zlink::monitor_event::connection_ready);

    assert (!server_monitor.recv (zlink::recv_flags_t::dontwait));
    assert (!client_monitor.recv (zlink::recv_flags_t::dontwait));

    server.bind ("tcp://127.0.0.1:0");
    const std::string endpoint = server.options ().last_endpoint ();
    assert (!endpoint.empty ());
    client.connect (endpoint);

    const std::optional<zlink::monitor_event_t> server_event = server_monitor.recv ();
    const std::optional<zlink::monitor_event_t> client_event = client_monitor.recv ();
    assert (server_event.has_value ());
    assert (client_event.has_value ());
    assert (server_event->event == zlink::monitor_event::connection_ready);
    assert (client_event->event == zlink::monitor_event::connection_ready);
    assert (!server_monitor.recv (zlink::recv_flags_t::dontwait));
    assert (!client_monitor.recv (zlink::recv_flags_t::dontwait));

    std::printf ("[monitor/recv] recv: \"connection-ready\" -> recv(non_blocking): empty\n");
    return 0;
}

File: bindings/javascript/samples/monitor_recv_sample.js
// SPDX-License-Identifier: MPL-2.0

'use strict';

const assert = require('node:assert/strict');
const { once } = require('node:events');
const net = require('node:net');
const zlink = require('@zlink-systems/zlink');

async function reservePort() {
  const srv = net.createServer();
  srv.listen(0, '127.0.0.1');
  await once(srv, 'listening');
  const { port } = srv.address();
  await new Promise((resolve, reject) => srv.close((error) => error ? reject(error) : resolve()));
  return port;
}

async function main() {
  const port = await reservePort();
  const endpoint = `tcp://127.0.0.1:${port}`;
  const ctx = zlink.createContext();
  const server = zlink.createPairSocket(ctx);
  const client = zlink.createPairSocket(ctx);
  const serverMonitor = server.monitorOpen([zlink.MonitorEventType.ConnectionReady]);
  const clientMonitor = client.monitorOpen([zlink.MonitorEventType.ConnectionReady]);

  try {
    server.bind(endpoint);
    client.connect(endpoint);

    const serverEvent = serverMonitor.recv();
    const clientEvent = clientMonitor.recv();
    assert.equal(serverEvent.event, zlink.MonitorEventType.ConnectionReady);
    assert.equal(clientEvent.event, zlink.MonitorEventType.ConnectionReady);
    console.log('[monitor/recv] recv: "connection-ready"');
  } finally {
    clientMonitor.close();
    serverMonitor.close();
    client.close();
    server.close();
    ctx.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
